### PR TITLE
bt-216: Create contact candidate modal

### LINKED ISSUE #216: Add Contact candidate modal
**User story:**
As an employer, I want to be able to contact a candidate, so that I can send information about a potential job.

**Acceptance criteria:**

- When an employer clicks on the “Contact candidate” button on the Candidat's closed page, the modal window should appear.
- The modal window must include input fields for “Add link to the vacancy”, “Write a message”
- The Add link to the vacancy”, “Write a message” fields should have the placeholder
- “Add link to the vacancy” field accepts only links in the format of: www. OR http:// OR https:// + >0 symbols + "." + >0 symbols
- “Add link to the vacancy” field length should be 5-250 chars and accept only: A-Z, a-z, 0-9, special chars
- Employer should be able to add more links, by clicking on “+Add more link”
- Employer should be able to add up to 5 links
- Employer should be able to delete added links by clicking on the delete icon
- “Write a message” field must be required and have “*”
- “Write a message” field length should be 10-5000 chars and accept A-Z, a-z, 0-9, special chars, spaces
- The modal window must include “Save as template” checkbox
- When the employer chooses the “Save as template” checkbox, the field for adding the template name should appear
- When the employer opens the “Contact candidate” modal window next time, it should be a list of saved templates for the user to choose from, a if template was saved
- The template name field should have the "Template name" placeholder
- Employer should be able to edit the name of template and delete template
- There should be edit and delete buttons right of the template name field
- The modal window must include the “Send message” button to send information to the talent

Image

Image
Image

### PR DESCRIPTION
**Pull request by ptarho**
I have created contact candidate modal component and candidate page in order to test this component (we can keep it, just fill with candidate page layout later). Also data from form is validated and for now log into a console. If user select save as a template message template will be saved in redux (I wasn't sure where to create slice, maybe better fit into chat bundle?)
When user will open modal component next time he will see a list of templates to use, by clicking on it the massage field value will change
To test this component just enter `/contact-talent` route

Screenshot_3
Screenshot_7
Screenshot_5
Screenshot_6

### CODE REVIEW

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+                                    name={'templateName'}
+                                    placeholder="Template name"
+                                />
+                            )}
+                        </FormControl>
+
+                        {
+                            // eslint-disable-next-line @typescript-eslint/no-magic-numbers
+                            messageTemplates.length > 0 && (
+                                <Grid>
```
Comment (lavandosya) [inline frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]: You can move it to constants.tsx

Comment (lavandosya) [reply frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]: Also think we should disable add more than 5 links, not only in validation

Comment (lavandosya) [reply frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]: And link length should be 5-250 chars, not to 7-50 chars

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    const { fields, append, remove } = useFieldArray({
+        name: 'links',
+        control,
+    });
+
+    const addLink = useCallback(() => {
+        append({
+            value: '',
+        });
+    }, [append]);
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]: I think we don't need use useCallback hook for this functions

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+    const removeLink = useCallback(
+        (index: number) => (): void => {
+            remove(index);
+        },
+        [remove],
+    );
+
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx:76]: The same

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    const applyTemplate = useCallback(
+        (message: string) => (): void => {
+            setValue('message', message);
+        },
+        [setValue],
+    );
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx:82]: Don't use useCallback in every function (https://dmitripavlutin.com/react-usecallback/)

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+    const handleFormSubmit = useCallback(
+        (event_: React.BaseSyntheticEvent): void => {
+            void handleSubmit(onSubmit)(event_);
+        },
+        [handleSubmit, onSubmit],
+    );
+
+    return (
+        isOpen && (
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]: Why we need isOpen && ... if we pass isOpen to Modal component ?

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+                                    control={control}
+                                    errors={errors}
+                                    name={'templateName'}
+                                    placeholder="Template name"
+                                />
+                            )}
+                        </FormControl>
+
+                        {
+                            // eslint-disable-next-line @typescript-eslint/no-magic-numbers
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]: Don't use eslint-disable. Just don't use magic numbers

File: frontend/src/bundles/candidate/pages/candidate.tsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    }, []);
+    const openModal1 = useCallback(() => {
+        setIsModal1Open(true);
+    }, []);
+    const closeModal2 = useCallback(() => {
+        setIsModal2Open(false);
+    }, []);
+    const openModal2 = useCallback(() => {
+        setIsModal2Open(true);
+    }, []);
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/candidate/pages/candidate.tsx]: Code duplication. In this case good practice to create hook for example useVisibility to manage modal isOpen state.

File: frontend/src/index.tsx [hunk 8]
```diff
@@ -43,6 +44,18 @@ createRoot(document.querySelector('#root') as HTMLElement).render(
...
+                                        path: AppRoute.SIGN_IN,
+                                        element: <Auth />,
+                                    },
+                                    {
+                                        path: AppRoute.SIGN_UP,
+                                        element: <Auth />,
+                                    },
+                                    {
+                                        path: AppRoute.CONTACT_TALENT, // Remove after testing
+                                        element: <Candidate />,
```
Comment (MaksymGavatuk) [inline frontend/src/index.tsx]: Need to remove before merge ?

**Review by MaksymGavatuk — CHANGES_REQUESTED** [review]

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+        defaultValues: { templates: messageTemplates },
+    });
+
+    return (
+        <Modal
+            isOpen={isOpen}
+            onClose={onClose}
+            headerLabel="Contact candidate"
+        >
+            <form className={styles.form} onSubmit={handleFormSubmit}>
```
Comment (nikita-remeslov) [inline frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx:143]: FormControl

File: frontend/src/bundles/candidate/components/candidate-modal/styles.module.scss [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    }
+}
+
+.addLink {
+    margin-top: 5px;
+    color: var(--vivid);
+    font-weight: 400;
+    font-size: 14px;
+
+    span {
```
Comment (nikita-remeslov) [inline frontend/src/bundles/candidate/components/candidate-modal/styles.module.scss]: B10 - https://github.com/BinaryStudioAcademy/quality-criteria/blob/production/src/javascript.md

File: frontend/src/bundles/candidate/components/candidate-modal/styles.module.scss [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    width: 100%;
+    max-width: 330px;
+}
+
+.message {
+    padding: 10px 15px;
+    border-color: var(--botticelli);
+}
+
+.messageLabel span:last-child {
```
Comment (nikita-remeslov) [inline frontend/src/bundles/candidate/components/candidate-modal/styles.module.scss]: https://github.com/BinaryStudioAcademy/quality-criteria/blob/production/src/javascript.md

File: frontend/src/bundles/candidate/components/message-template/message-template.tsx [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+    );
+
+    const removeTemplate = useCallback(
+        (templateName: string) => (): void => {
+            void dispatch(candidateActions.removeMessageTemplate(templateName));
+        },
+        [dispatch],
+    );
+
+    return isEdit ? (
```
Comment (nikita-remeslov) [inline frontend/src/bundles/candidate/components/message-template/message-template.tsx:96]: why do we need this logic?

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]
you should add an error state for inputs
image

File: frontend/src/bundles/candidate/components/message-template/message-template.tsx (hunk 12 above)
Comment (ptarho) [reply frontend/src/bundles/candidate/components/message-template/message-template.tsx:96]: In order to change template name. When the user click on change icon we want to replace button with input to let user type new name. And  also change buttons icons to confirm or decline editing

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx (hunk 6 above)
Comment (ptarho) [reply frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]: fixed

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx (hunk 5 above)
Comment (ptarho) [reply frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]: fixed

File: frontend/src/bundles/candidate/pages/candidate.tsx (hunk 7 above)
Comment (ptarho) [reply frontend/src/bundles/candidate/pages/candidate.tsx]: This code is only for demonstration/testing of this component. Do I need to change this part if we will remove it before merging?

File: frontend/src/index.tsx (hunk 8 above)
Comment (ptarho) [reply frontend/src/index.tsx]: Yes

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx (hunk 9 above)
Comment (ptarho) [reply frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx:143]: @nikita-remeslov why do we have to use FormControl? Our sign-in/sign-up pages use default form tag

File: frontend/src/bundles/candidate/components/candidate-modal/styles.module.scss (hunk 10 above)
Comment (ptarho) [reply frontend/src/bundles/candidate/components/candidate-modal/styles.module.scss]: Fixed

File: frontend/src/bundles/candidate/components/candidate-modal/styles.module.scss (hunk 11 above)
Comment (ptarho) [reply frontend/src/bundles/candidate/components/candidate-modal/styles.module.scss]: Fixed this request and also input error state

File: frontend/src/assets/css/variables.scss [hunk 13]
```diff
@@ -38,9 +41,6 @@
     // Porcelain
     --porcelain: #f5f6f7;
 
-    // Sunglow
-    --sunglow: #ffd231;
-
```
Comment (nika-sirak) [inline frontend/src/assets/css/variables.scss]: Why do we delete this variable?

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+                            return (
+                                <Grid key={index}>
+                                    <Grid
+                                        className={styles.linkWrapper}
+                                        key={field.id}
+                                    >
+                                        <Input
+                                            className={getValidClassNames(
+                                                styles.link,
+                                            )}
```
Comment (nika-sirak) [inline frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]:
```suggestion
                                            className={styles.link}
```

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+                        />
+                        {errors.links && (
+                            <FormHelperText className={styles.error}>
+                                {errors.links.message}
+                            </FormHelperText>
+                        )}
+                    </FormControl>
+
+                    <FormControl
+                        className={getValidClassNames(styles.formControl)}
```
Comment (nika-sirak) [inline frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]:
```suggestion
                        className={styles.formControl}
```

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+                                styles.messageLabel,
+                            )}
+                            required
+                        >
+                            <Typography variant="label">
+                                Write a message
+                            </Typography>
+                        </FormLabel>
+                        <Textarea
+                            className={getValidClassNames(styles.message)}
```
Comment (nika-sirak) [inline frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]:
```suggestion
                            className={styles.message}
```

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx [hunk 17]
```diff
@@ -0,0 +1,297 @@
+import {
+    Add as AddIcon,
+    DeleteOutline as DeleteIcon,
+} from '@mui/icons-material';
+import { FormHelperText } from '@mui/material';
```
Comment (nika-sirak) [inline frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]: We have component. Import from '~/bundles/common/components/components.js';

File: frontend/src/bundles/candidate/components/message-template/message-template.tsx [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+                    placeholder="Template name"
+                    control={control}
+                    errors={errors}
+                    name={name}
+                    value={editedName}
+                    onChange={handleNameChange}
+                />
+                <Grid className={styles.buttonsContainer}>
+                    <Button
+                        className={getValidClassNames(styles.button)}
```
Comment (nika-sirak) [inline frontend/src/bundles/candidate/components/message-template/message-template.tsx:116]:
```suggestion
                        className={styles.button}
```

File: frontend/src/bundles/candidate/components/message-template/message-template.tsx [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+                        label=""
+                        onClick={confirmTemplateEdit(template.name)}
+                        variant="outlined"
+                        endIcon={<CheckIcon className={styles.buttonIcon} />}
+                    />
+                    <Button
+                        className={getValidClassNames(styles.button)}
```
Comment (nika-sirak) [inline frontend/src/bundles/candidate/components/message-template/message-template.tsx:123]:
```suggestion
                        className={styles.button}
```

File: frontend/src/bundles/candidate/components/message-template/message-template.tsx [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+                    styles.templateButton,
+                    styles.template,
+                )}
+                onClick={applyTemplate(template.message)}
+                variant="text"
+                label={template.name}
+            />
+            <Grid className={styles.buttonsContainer}>
+                <Button
+                    className={getValidClassNames(styles.button)}
```
Comment (nika-sirak) [inline frontend/src/bundles/candidate/components/message-template/message-template.tsx:151]:
```suggestion
                    className={styles.button}
```

File: frontend/src/bundles/candidate/components/message-template/message-template.tsx [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+                    label=""
+                    onClick={editTemplate}
+                    variant="outlined"
+                    endIcon={<EditIcon className={styles.buttonIcon} />}
+                />
+                <Button
+                    className={getValidClassNames(styles.button)}
```
Comment (nika-sirak) [inline frontend/src/bundles/candidate/components/message-template/message-template.tsx:158]:
```suggestion
                    className={styles.button}
```

File: frontend/src/assets/css/variables.scss (hunk 13 above)
Comment (ptarho) [reply frontend/src/assets/css/variables.scss]: There was a duplication error when I deleted it, but looks like we already fixed in development branch, so I will return it

File: frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx (hunk 17 above)
Comment (ptarho) [reply frontend/src/bundles/candidate/components/candidate-modal/candidate-modal.tsx]: fixed

**Comment by nikita-remeslov** [comment]
Conflicts

File: shared/src/bundles/candidate/validation-schemas/contact-modal.validation-schema.ts [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+            'string.required':
+                ContactCandidateValidationMessage.MESSAGE_REQUIRED,
+            'string.empty': ContactCandidateValidationMessage.MESSAGE_REQUIRED,
+        }),
+
+    isSaveTemplate: joi.boolean().required(),
+
+    templateName: joi.string().when('isSaveTemplate', {
+        is: true,
+        // eslint-disable-next-line unicorn/no-thenable
```
Comment (nikita-remeslov) [inline shared/src/bundles/candidate/validation-schemas/contact-modal.validation-schema.ts:61]: ?

**Review by nikita-remeslov — APPROVED** [review]

**Review by lavandosya — APPROVED** [review]

**Review by nika-sirak — APPROVED** [review]

**Pull request merged**